File: app/src/main/java/com/ch/yoon/kakao/pay/imagesearch/repository/ImageRepository.java
package com.ch.yoon.kakao.pay.imagesearch.repository;

import androidx.annotation.NonNull;

import com.ch.yoon.kakao.pay.imagesearch.repository.local.room.entity.SearchLog;
import com.ch.yoon.kakao.pay.imagesearch.repository.model.imagesearch.request.ImageSearchRequest;
import com.ch.yoon.kakao.pay.imagesearch.repository.model.imagesearch.response.ImageSearchResponse;

import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Single;

/**
 * Creator : ch-yoon
 * Date : 2019-08-01.
 */
public interface ImageRepository {

    @NonNull
    Single<ImageSearchResponse> requestImageList(@NonNull ImageSearchRequest imageSearchRequest);

    @NonNull
    Completable deleteSearchLog(@NonNull String keyword);

    @NonNull
    Single<List<SearchLog>> requestSearchLogList();

    @NonNull
    Single<SearchLog> insertOrUpdateSearchLog(@NonNull String keyword);

}

File: app/src/main/java/com/ch/yoon/imagesearch/data/repository/ImageRepository.kt
package com.ch.yoon.imagesearch.data.repository

import com.ch.yoon.imagesearch.data.remote.kakao.request.ImageSearchRequest
import com.ch.yoon.imagesearch.data.repository.model.ImageSearchResponse
import com.ch.yoon.imagesearch.data.repository.model.SearchLogModel
import io.reactivex.Completable
import io.reactivex.Single

/**
 * Creator : ch-yoon
 * Date : 2019-10-28
 **/
interface ImageRepository {

    fun requestImageList(imageSearchRequest: ImageSearchRequest): Single<ImageSearchResponse>

    fun deleteSearchLog(searchLogModel: SearchLogModel): Completable

    fun requestSearchLogList(): Single<List<SearchLogModel>>

    fun insertOrUpdateSearchLog(keyword: String): Single<SearchLogModel>

    fun deleteAllSearchLog(): Completable

}